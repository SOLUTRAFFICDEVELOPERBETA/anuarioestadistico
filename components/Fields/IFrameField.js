import styled from '@emotion/styled';
import {
    Box,
    Collapse,
    Grid,
    IconButton,
    InputAdornment,
    InputBase,
    Popover,
    Tooltip
} from '@material-ui/core';
import React from 'react';
import PropTypes from 'prop-types';
import { DeleteForever, DragHandle, Http, Send } from '@material-ui/icons';
import { grey } from '@material-ui/core/colors';
import { DragPreviewImage, useDrag } from 'react-dnd';
import AlertContext from '../../contexts/alert';

// Componente contenedor del IFrame
const IFrameContainer = styled.div`
    width: 100%;
    padding: 16px;
    transition: background-color 0.5s ease, z-index 0.5s ease, box-shadow 0.5s ease;
    display: grid;
    grid-template-columns: 0px auto;

    :focus-within {
        background-color: #eee;
        z-index: 5;
        -webkit-box-shadow: 0px 0px 5px 0px rgba(0, 0, 0, 0.75);
        -moz-box-shadow: 0px 0px 5px 0px rgba(0, 0, 0, 0.75);
        box-shadow: 0px 0px 5px 0px rgba(0, 0, 0, 0.75);
        grid-template-columns: 0px auto;
        transition: 0.5s;
    }

    :hover {
        grid-template-columns: 30px auto;
    }

    .dragger {
        align-self: center;
        justify-self: center;
        width: 24px;
        height: 24px;
        visibility: hidden;
        opacity: 0;
        transition: visibility 0s, opacity 0.5s linear;
    }

    :hover .dragger {
        visibility: visible;
        opacity: 1;
    }

    .dragger:hover {
        cursor: move;
    }

    .iframe-ref {
        padding: 4px;
        border: 1px solid var(--primary);
        transition: border 0.5s ease;
    }

    .iframe-ref[data-ref=''] {
        border: 1px dashed var(--primary);
    }

    iframe {
        width: 100%;
        height: 25rem;
        transition: height 0.5s ease;
    }

    iframe[src=''] {
        display: none;
        height: 0rem;
    }
`;

/**
 * Método para validar el link de powerbi
 * @param {String} link Link a revisar
 */
const validateLink = (link) => link.includes('https://app.powerbi.com/');

/**
 * Componente para visualizar un IFrame de PowerBI
 * @param {{ id: string, value: string, onChange: () => void, onDelete: () => void }} props Propiedades del componente
 */
const IFrameField = ({ id, value, onChange, onDelete }) => {
    const [ref, setRef] = React.useState(value);
    const { showMessage } = React.useContext(AlertContext);
    const [anchorEl, setAnchorEl] = React.useState(null);
    const [{ isDragging }, drag, preview] = useDrag({
        item: { type: 'ITEM', id },
        collect: (monitor) => ({
            isDragging: !!monitor.isDragging()
        })
    });

    /**
     * Método para abrir el popover
     * @param {Event} event Evento de click de la celda
     */
    const handleClick = (event) => {
        event.preventDefault();
        setAnchorEl(event.currentTarget);
    };

    // Método para cerrar el popover
    const handleClose = () => setAnchorEl(null);

    /**
     * Método para cambiar la referencia
     * @param {Event} event Evento del Input
     */
    const handleChange = ({ target: { value } }) => setRef(value);

    /**
     * Método para guardar la referencia introducida
     * @param {Event} event Evento de Submit
     */
    const handleSetUrl = (event) => {
        event.preventDefault();
        if (validateLink(ref)) {
            onChange({ value: ref });
        } else {
            showMessage('El vinculo ingresado no es valido');
        }
    };

    const open = Boolean(anchorEl);

    if (isDragging) return null;

    return (
        <React.Fragment>
            <DragPreviewImage connect={preview} src="/static/img/DragPreview.png" />
            <IFrameContainer ref={drag} onContextMenu={handleClick}>
                <span className="dragger">
                    <DragHandle fontSize="small" style={{ fill: grey[500] }} />
                </span>
                <div className="iframe-ref" data-ref={value} data-validate={ref}>
                    <form onSubmit={handleSetUrl}>
                        <InputBase
                            defaultValue={value}
                            value={ref}
                            name="url"
                            type="url"
                            size="small"
                            margin="dense"
                            required
                            fullWidth
                            placeholder="Introduce la dirección del documento"
                            onChange={handleChange}
                            startAdornment={
                                <InputAdornment position="start">
                                    <Http color="primary" />
                                </InputAdornment>
                            }
                            endAdornment={
                                <InputAdornment position="end">
                                    <Tooltip title="Cambiar">
                                        <IconButton size="small" type="submit">
                                            <Send color="primary" />
                                        </IconButton>
                                    </Tooltip>
                                </InputAdornment>
                            }
                        />
                    </form>
                    <Collapse in={value !== ''}>
                        <iframe id={`iframe-view-${id}`} title={`${id}`} src={value} />
                    </Collapse>
                </div>
            </IFrameContainer>
            {open && (
                <Popover
                    id={`popover-open-${id}`}
                    open={open}
                    anchorEl={anchorEl}
                    onClose={handleClose}
                    anchorOrigin={{
                        vertical: 'top',
                        horizontal: 'left'
                    }}
                    transformOrigin={{
                        vertical: 'bottom',
                        horizontal: 'left'
                    }}>
                    <Box padding={1}>
                        <Grid container spacing={1} alignContent="center" alignItems="center">
                            <Grid item>
                                <Tooltip title="Borrar" interactive>
                                    <IconButton size="small" onClick={() => onDelete(id)}>
                                        <DeleteForever color="error" />
                                    </IconButton>
                                </Tooltip>
                            </Grid>
                        </Grid>
                    </Box>
                </Popover>
            )}
        </React.Fragment>
    );
};

IFrameField.propTypes = {
    id: PropTypes.string.isRequired,
    value: PropTypes.object,
    onChange: PropTypes.func.isRequired,
    onDelete: PropTypes.func.isRequired
};

export default IFrameField;
