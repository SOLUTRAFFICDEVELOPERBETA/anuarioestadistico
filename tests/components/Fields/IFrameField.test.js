import '@testing-library/jest-dom';
import { act } from '@testing-library/react';
import { mount } from 'enzyme';
import React from 'react';
import { DndProvider } from 'react-dnd'
import { HTML5Backend } from 'react-dnd-html5-backend'
import IFrameField from '../../../components/Fields/IFrameField';
import AlertContext from '../../../contexts/alert';


describe('Pruebas en <IFrameField/>', () => {

    const onDelete = jest.fn();
    const onChange = jest.fn();
    const showMessage = jest.fn();
    const value = {
        ref: 'https://app.powerbi.com/view?r=eyJrIjoiNmQ3ODYwOTUtZTBiMC00NWQxLWJjYTYtMGQ5ZjI4MDkwOGJlIiwidCI6IjMzZmE2OTA3LTIxYzYtNDEyNi1hYTY2LWJjYmVkOTUzNjFiYiIsImMiOjR9'
    }

    test('debe mostrarse correctamente', () => {
        const wrapper = mount(
            <DndProvider backend={HTML5Backend}>
                <AlertContext.Provider value={{
                    showMessage
                }}>
                    <IFrameField
                        id="iframent-id"
                        value={value}
                        onDelete={onDelete}
                        onChange={onChange}
                    />
                </AlertContext.Provider>
            </DndProvider>
        )
        expect(wrapper).toMatchSnapshot()
    })

    // test('debe de llamar la función onChange()', () => {
    //     act(() => {
    //         wrapper.find('input').prop('onChange')({ target: { value: value } })
    //     })
    //     wrapper.find('form').prop('onSubmit')({ preventDefault() { } });

    //     expect(onChange).toHaveBeenCalled();
    //     expect(onChange).toHaveBeenCalledTimes(1);
    //     expect(onChange).toHaveBeenCalledWith(value);
    // })

    test('debe llamar la función, onDelete()', () => {
        // const wrapper = mount(
        //     <DndProvider backend={HTML5Backend}>
        //         <AlertContext.Provider value={{
        //             showMessage
        //         }}>
        //             <IFrameField
        //                 id="iframent-id"
        //                 value={value}
        //                 onDelete={onDelete}
        //                 onChange={onChange}
        //             />
        //         </AlertContext.Provider>
        //     </DndProvider>
        // )
        // act(() => {
        //     wrapper.find('div').at(0).prop('onContextMenu')({ preventDefault() { }, currentTarget() { } });
        // })
        // expect(wrapper).toMatchSnapshot();
        // console.log(wrapper.html());

    })


})