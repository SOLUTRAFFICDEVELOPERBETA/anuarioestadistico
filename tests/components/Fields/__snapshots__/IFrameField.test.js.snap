// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <IFrameField/> debe mostrarse correctamente 1`] = `
Array [
  "",
  <div
    className="css-12ru47l-IFrameContainer e7iy8zi0"
    onContextMenu={[Function]}
  >
    <span
      className="dragger"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
        focusable="false"
        style={
          Object {
            "fill": "#9e9e9e",
          }
        }
        viewBox="0 0 24 24"
      >
        <path
          d="M20 9H4v2h16V9zM4 15h16v-2H4v2z"
        />
      </svg>
    </span>
    <div
      className="iframe-ref"
      data-ref={
        Object {
          "ref": "https://app.powerbi.com/view?r=eyJrIjoiNmQ3ODYwOTUtZTBiMC00NWQxLWJjYTYtMGQ5ZjI4MDkwOGJlIiwidCI6IjMzZmE2OTA3LTIxYzYtNDEyNi1hYTY2LWJjYmVkOTUzNjFiYiIsImMiOjR9",
        }
      }
      data-validate={
        Object {
          "ref": "https://app.powerbi.com/view?r=eyJrIjoiNmQ3ODYwOTUtZTBiMC00NWQxLWJjYTYtMGQ5ZjI4MDkwOGJlIiwidCI6IjMzZmE2OTA3LTIxYzYtNDEyNi1hYTY2LWJjYmVkOTUzNjFiYiIsImMiOjR9",
        }
      }
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="MuiInputBase-root MuiInputBase-fullWidth MuiInputBase-adornedStart MuiInputBase-adornedEnd MuiInputBase-marginDense"
          onClick={[Function]}
          required={true}
          size="small"
        >
          <div
            className="MuiInputAdornment-root MuiInputAdornment-positionStart"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-colorPrimary"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M4.5 11h-2V9H1v6h1.5v-2.5h2V15H6V9H4.5v2zm2.5-.5h1.5V15H10v-4.5h1.5V9H7v1.5zm5.5 0H14V15h1.5v-4.5H17V9h-4.5v1.5zm9-1.5H18v6h1.5v-2h2c.8 0 1.5-.7 1.5-1.5v-1c0-.8-.7-1.5-1.5-1.5zm0 2.5h-2v-1h2v1z"
              />
            </svg>
          </div>
          <input
            className="MuiInputBase-input MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd MuiInputBase-inputMarginDense"
            name="url"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Introduce la direcciÃ³n del documento"
            required={true}
            type="url"
            value={
              Object {
                "ref": "https://app.powerbi.com/view?r=eyJrIjoiNmQ3ODYwOTUtZTBiMC00NWQxLWJjYTYtMGQ5ZjI4MDkwOGJlIiwidCI6IjMzZmE2OTA3LTIxYzYtNDEyNi1hYTY2LWJjYmVkOTUzNjFiYiIsImMiOjR9",
              }
            }
          />
          <div
            className="MuiInputAdornment-root MuiInputAdornment-positionEnd"
          >
            Array [
              <button
                aria-describedby={null}
                className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
                disabled={false}
                onBlur={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseOver={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                title="Cambiar"
                type="submit"
              >
                <span
                  className="MuiIconButton-label"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root MuiSvgIcon-colorPrimary"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M2.01 21L23 12 2.01 3 2 10l15 2-15 2z"
                    />
                  </svg>
                </span>
                <span
                  className="MuiTouchRipple-root"
                />
              </button>,
              "",
            ]
          </div>
        </div>
      </form>
      <div
        className="MuiCollapse-container MuiCollapse-entered"
        style={
          Object {
            "minHeight": "0px",
          }
        }
      >
        <div
          className="MuiCollapse-wrapper"
        >
          <div
            className="MuiCollapse-wrapperInner"
          >
            <iframe
              id="iframe-view-iframent-id"
              src={
                Object {
                  "ref": "https://app.powerbi.com/view?r=eyJrIjoiNmQ3ODYwOTUtZTBiMC00NWQxLWJjYTYtMGQ5ZjI4MDkwOGJlIiwidCI6IjMzZmE2OTA3LTIxYzYtNDEyNi1hYTY2LWJjYmVkOTUzNjFiYiIsImMiOjR9",
                }
              }
              title="iframent-id"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
